man/addftinfo.man
.\" -*- nroff -*-
.TH ADDFTINFO @MAN1EXT@ "@MDATE@" "Groff Version @VERSION@"
.SH NAME
addftinfo \- add information to troff font files for use with groff
.SH SYNOPSIS
.B addftinfo
[
.BI \- param\ value\fR.\|.\|.
]
.I res
.I unitwidth
.I font
.SH DESCRIPTION
.B addftinfo
reads a troff font file
and adds some additional font-metric information
that is used by the groff system.
The font file with the information added is written on the
standard output.
The information added is guessed using 
some parametric information about the font
and assumptions
about the traditional troff names for characters.
The main information added is the heights and depths of characters.
The
.I res
and
.I unitwidth
arguments should be the same as the corresponding parameters
in the DESC file;
.I font
is the name of the file describing the font;
if
.I font
ends with
.B I
the font will be assumed to be italic.
.SH OPTIONS
Each of the options changes one of the parameters that is used
to derive the heights and depths.
Like the existing quantities in the font
file, each
.I value
is in
.RI inches/ res
for a font whose point size is
.IR unitwidth .
.I param
must be one of:
.TP
.B x-height
The height of lowercase letters without ascenders such as x.
.TP
.B fig-height
The height of figures (digits).
.TP
.B asc-height
The height of characters with ascenders, such as b, d or l.
.TP
.B body-height
The height of characters such as parentheses.
.TP
.B cap-height
The height of uppercase letters such as A.
.TP
.B comma-depth
The depth of a comma.
.TP
.B desc-depth
The depth of characters with descenders, such as p,q, or y.
.TP
.B body-depth
The depth of characters such as parentheses.
.LP
.B addftinfo
makes no attempt to use the specified parameters to guess
the unspecified parameters.
If a parameter is not specified the default will be used.
The defaults are chosen to have the reasonable values for
a Times font.
.SH "SEE ALSO"
.BR font (5)
.BR groff_font (@MAN5EXT@),
.BR groff (@MAN1EXT@),
.I "Groff Character Names"


NEWS
This file describes recent user-visible changes in groff.  Bug fixes
are not described.  There are more details in the man pages.

Groff
-----

o `-mFOO' will now search first for `FOO.tmac' and then for `tmac.FOO'.  The
  old behaviour has been changed to overcome problems with platforms which
  have an 8+3 file name limit, and platforms which have other versions of
  troff installed also. Additionally, all macro files have been renamed
  using the latter scheme to avoid 8+3 name clashes.

o The new environment variable GROFF_BIN_PATH will be checked for programs
  groff is calling (preprocessors, troff, and output devices) before PATH.
  If not set, it defaults to the directory where the groff binary is
  located.  Previously, it was PATH only.  The nroff script will only use
  GROFF_BIN_PATH to find the groff binary but passes both the GROFF_BIN_PATH
  and PATH environment variable to groff.

Troff
-----

o The mdoc package has been completely rewritten, using the full power of
  GNU troff to remove limitations of Unix troff (which is no longer
  supported).  Most important changes are:

  . No argument limit.
  . Almost all macros are parsed and callable (if it makes sense).
  . `.Lb': prints library names
  . `.Nm <punctuation>' now works as expected; `.Nm "" <punctuation>' has
    been withdrawn.
  . Updated `.St' command
  . `.Fx': prints FreeBSD
  . `.Ox': prints OpenBSD
  . `.Bsx': prints BSD/OS
  . `.Brq', `.Bro', `.Brc': brace enclosure macros
  . `.Bd -centered': center lines
  . `.Bl -xwidth <string>': interpret <string> and use the resulting width
  . support for double-sided printing (-rD1 command line switch)
  . support for 11pt and 12pt document sizes (-rS11, -rS12 command line
    switches)

  `groff_mdoc.man' replaces `groff_mdoc.samples.man'; it now completely
  documents the mdoc package.

  Great care has been taken to assure backwards compatibility.  If you
  encounter any abnormal results, please report them to bug-groff@gnu.org.

o A new command line option for the `man' macros (similar to the `mdoc'
  package has been implemented: `-rcR=1' (now the default in nroff mode) will
  produce one single, very long page instead of multiple pages.  `-rcR=0'
  will deactivate it.

o The `return' request has been added to return immediately from a macro.

o A new request `nop' (no operation) has been added which is similar to
  `if 1'.  For example,

    .if t \{\
    Hallo!
    .\}

  can now be written as

    .if t \{\
    .  nop Hallo!
    .\}

o `box' and `boxa' are two new requests which behave similarly to `di' and
  `da' but don't include a partially filled line (which is restored after
  ending the diversion).

o The `asciify' request has been extended to `unformat' space characters
  and some other escape sequences also.

  `\ ' will no longer be unformatted as a space but remains an unpaddable,
  unbreakable space character.

o The new `unformat' request is similar to `asciify' but only handles space
  characters and tabs specially if the diversion is reread, retaining font
  information.  This makes it possible to reformat diversions; for example
  the following

    .ll 3i
    .
    a01 a02 a03 a04 a05 a06 a07 a08 a09 a10.
    .
    .box box1
    .ev 1
    .nf
    \f[B]b01 b02 b03 b04 b05 b06 b07 b08 b09 b10.\f[P]
    .br
    .ev
    .box
    .
    c01 c02 c03 c04 c05 c06 c07 c08 c09 c10.
    .
    .unformat box1
    .box1

  gives

    a01  a02  a03  a04 a05 a06 a07
    a08 a09 a10.  c01 c02 c03  c04
    c05  c06 c07 c08 c09 c10.  b01
    b02 b03 b04 b05  b06  b07  b08
    b09 b10.

  Without the `unformat' request, space characters are converted to word
  space nodes which are no longer stretchable, and the result would be

    a01  a02  a03  a04 a05 a06 a07
    a08 a09 a10.  c01 c02 c03  c04
    c05  c06 c07 c08 c09 c10.  b01
    b02 b03 b04 b05 b06 b07 b08
    b09 b10.

o The new request `linetabs' controls the `line-tabs' mode.  In line-tabs
  mode, tab distances are computed relative to the (current) output line.
  Otherwise they are taken relative to the input line.  For example, the
  following

    .ds x a\t\c
    .ds y b\t\c
    .ds z c
    .ta 1i 3i
    \*x
    \*y
    \*z

  yields

    a         b         c

  In line-tabs mode, the same code gives

    a         b                   c

  The new read-only number register `.linetabs' returns 1 if in line-tabs
  mode, and 0 otherwise.

o Two new requests `tm1' and `tmc' have been added to improve writing
  messages to the terminal.  `tm1' is similar to `tm' but allows leading
  whitespace.  `tmc' is similar to `tm1' but doesn't emit a final newline.

o A new request `dei' (define indirect) has been added.  The first and
  second parameter of `dei' are taken from string registers rather than
  directly; this very special request is needed to make `trace.tmac'
  independent from the escape character (which might even be disabled).

o It is now possible to save and restore the escape character with two new
  requests `ecs' and `ecr'.

o The new escape sequence \B'...' is an analogon to `\A': If the string
  within the delimiters is a valid numeric expression, return character `1',
  and `0' otherwise.

o The new escape sequence `\:' inserts a zero-width break point.  This is
  similar to `\%' but without a soft hyphen character.

o The `tr' request can now map characters onto `\~'.

o Calling the `fam' request without an argument switches back to the
  previous font family.

o The new read-only register `.int' is set to a positive value if the last
  output line is interrupted (i.e., if it contains `\c').

o The `writem' request is not new, but hasn't been documented before.  This
  is similar to `write' but instead of a string the contents of a given
  macro or string is written to a stream.

o The read/write number register `hp' to get/set the current horizontal
  position relative to the input line isn't new but hasn't been documented
  properly before.


src/utils/addftinfo/addftinfo.cc
// -*- C++ -*-
/* Copyright (C) 1989-1992, 2000, 2001 Free Software Foundation, Inc.
     Written by James Clark (jjc@jclark.com)

This file is part of groff.

groff is free software; you can redistribute it and/or modify it under
the terms of the GNU General Public License as published by the Free
Software Foundation; either version 2, or (at your option) any later
version.

groff is distributed in the hope that it will be useful, but WITHOUT ANY
WARRANTY; without even the implied warranty of MERCHANTABILITY or
FITNESS FOR A PARTICULAR PURPOSE.  See the GNU General Public License
for more details.

You should have received a copy of the GNU General Public License along
with groff; see the file COPYING.  If not, write to the Free Software
Foundation, 59 Temple Place - Suite 330, Boston, MA 02111-1307, USA. */

#include <stdio.h>
#include <ctype.h>
#include <string.h>
#include <assert.h>
#include <stdlib.h>
#include <errno.h>
#include "lib.h"
#include "errarg.h"
#include "error.h"
#include "stringclass.h"
#include "cset.h"
#include "guess.h"

static void usage(FILE *stream);
static void usage();
static void version();
static void convert_font(const font_params &, FILE *, FILE *);

typedef int font_params::*param_t;

static struct {
  const char *name;
  param_t par;
} param_table[] = {
  { "x-height", &font_params::x_height },
  { "fig-height", &font_params::fig_height },
  { "asc-height", &font_params::asc_height },
  { "body-height", &font_params::body_height },
  { "cap-height", &font_params::cap_height },
  { "comma-depth", &font_params::comma_depth },
  { "desc-depth", &font_params::desc_depth },
  { "body-depth", &font_params::body_depth },
};

// These are all in thousandths of an em.
// These values are correct for PostScript Times Roman.

#define DEFAULT_X_HEIGHT 448
#define DEFAULT_FIG_HEIGHT 676
#define DEFAULT_ASC_HEIGHT 682
#define DEFAULT_BODY_HEIGHT 676
#define DEFAULT_CAP_HEIGHT 662
#define DEFAULT_COMMA_DEPTH 143
#define DEFAULT_DESC_DEPTH 217
#define DEFAULT_BODY_DEPTH 177

int main(int argc, char **argv)
{
  program_name = argv[0];
  int i;
  for (i = 1; i < argc; i++) {
    if (!strcmp(argv[i], "-v") || !strcmp(argv[i],"--version"))
      version();
    if (!strcmp(argv[i],"--help")) {
      usage(stdout);
      exit(0);
    }
  }
  if (argc < 4)
    usage();
  int resolution;
  if (sscanf(argv[argc-3], "%d", &resolution) != 1)
    usage();
  if (resolution <= 0)
    fatal("resolution must be > 0");
  int unitwidth;
  if (sscanf(argv[argc-2], "%d", &unitwidth) != 1)
    usage();
  if (unitwidth <= 0)
    fatal("unitwidth must be > 0");
  font_params param;
  const char *font = argv[argc-1];
  param.italic = (font[0] != '\0' && strchr(font, '\0')[-1] == 'I');
  param.em = (resolution*unitwidth)/72;
  param.x_height = DEFAULT_X_HEIGHT;
  param.fig_height = DEFAULT_FIG_HEIGHT;
  param.asc_height = DEFAULT_ASC_HEIGHT;
  param.body_height = DEFAULT_BODY_HEIGHT;
  param.cap_height = DEFAULT_CAP_HEIGHT;
  param.comma_depth = DEFAULT_COMMA_DEPTH;
  param.desc_depth = DEFAULT_DESC_DEPTH;
  param.body_depth = DEFAULT_BODY_DEPTH;
  for (i = 1; i < argc && argv[i][0] == '-'; i++) {
    if (argv[i][1] == '-' && argv[i][2] == '\0') {
      i++;
      break;
    }
    if (i + 1 >= argc)
      usage();
    size_t j;
    for (j = 0;; j++) {
      if (j >= sizeof(param_table)/sizeof(param_table[0]))
	fatal("parameter `%1' not recognized", argv[i] + 1);
      if (strcmp(param_table[j].name, argv[i] + 1) == 0)
	break;
    }
    if (sscanf(argv[i+1], "%d", &(param.*(param_table[j].par))) != 1)
      fatal("invalid argument `%1'", argv[i+1]);
    i++;
  }    
  if (argc - i != 3)
    usage();
  errno = 0;
  FILE *infp = fopen(font, "r");
  if (infp == 0)
    fatal("can't open `%1': %2", font, strerror(errno));
  convert_font(param, infp, stdout);
  return 0;
}

static void usage(FILE *stream)
{
  fprintf(stream, "usage: %s [-v] [-param value] ... "
		  "resolution unitwidth font\n",
	  program_name);
}
static void usage()
{
  usage(stderr);
  exit(1);
}

static void version()
{
  extern const char *Version_string;
  printf("GNU addftinfo (groff) version %s\n", Version_string);
  exit(0);
}

static int get_line(FILE *fp, string *p)
{
  int c;
  p->clear();
  while ((c = getc(fp)) != EOF) {
    *p += char(c);
    if (c == '\n')
      break;
  }
  return p->length() > 0;
}
  
static void convert_font(const font_params &param, FILE *infp, FILE *outfp)
{
  string s;
  while (get_line(infp, &s)) {
    put_string(s, outfp);
    if (s.length() >= 8
	&& strncmp(&s[0], "charset", 7))
      break;
  }
  while (get_line(infp, &s)) {
    s += '\0';
    string name;
    const char *p = s.contents();
    while (csspace(*p))
      p++;
    while (*p != '\0' && !csspace(*p))
      name += *p++;
    while (csspace(*p))
      p++;
    for (const char *q = s.contents(); q < p; q++)
      putc(*q, outfp);
    char *next;
    char_metric metric;
    metric.width = (int)strtol(p, &next, 10);
    if (next != p) {
      printf("%d", metric.width);
      p = next;
      metric.type = (int)strtol(p, &next, 10);
      if (next != p) {
	name += '\0';
	guess(name.contents(), param, &metric);
	if (metric.sk == 0) {
	  if (metric.left_ic == 0) {
	    if (metric.ic == 0) {
	      if (metric.depth == 0) {
		if (metric.height != 0)
		  printf(",%d", metric.height);
	      }
	      else
		printf(",%d,%d", metric.height, metric.depth);
	    }
	    else
	      printf(",%d,%d,%d", metric.height, metric.depth, metric.ic);
	  }
	  else
	    printf(",%d,%d,%d,%d", metric.height, metric.depth, metric.ic,
		   metric.left_ic);
	}
	else
	  printf(",%d,%d,%d,%d,%d", metric.height, metric.depth, metric.ic,
		 metric.left_ic, metric.sk);
      }
    }
    fputs(p, outfp);
  }
}

